claat/serve.go

```diff
@@ -0,0 +1,306 @@
+	"stacky":                        "PolymerElements/stacky#1.3.2",
+	"webcomponentsjs":               "WebComponents/webcomponentsjs#0.7.24",
+}
+
+// overrideComp maps a given spec, as described in fetchRepo doc comments to a
+// desired component name.
+var overrideComp = map[string]string{
+	"googlecodelabs/codelab-components": "google-codelab-elements",
+	"google/code-prettify":              "google-prettify",
+}
+
+// fetchRepo downloads a repo from github and unpacks it into basedir/dest,
+// calling itself recursively for every dependency found in the unpacked bower.json.
+// If the basedir/dest directory already exists, fetchRepo does nothing.
+//
+// The spec has the following format: "githubuser/repo#version".
+// Any non-alphanumeric characters are stripped from the version.
+// If version is missing, "master" is used instead.
+// The bower registry is not used for naming resolution.
+func fetchRepo(basedir, spec string) error {
+	var user, repo, path, comp, ver string
+	s := strings.Split(spec, "#")
+	path = s[0]
+	if len(s) > 1 {
+		ver = s[1]
+	}
+	ver = strings.Trim(ver, "^~ ")
+	if ver == "" {
+		ver = "master"
+	}
+	s = strings.Split(path, "/")
+	user = s[0]
+	if len(s) > 1 {
+		repo = s[1]
+	}
+	// Check exception map to see if we should override the component name.
+	comp = repo
+	if c, ok := overrideComp[path]; ok {
+		comp = c
+	}
+	// if repo already exists locally, return immediately, we're done.
+	if _, err := os.Stat(basedir + "/" + comp); err == nil {
+		return nil
+	}
+	zipFile := basedir + "/" + comp + ".zip"
+	url := "https://github.com/" + user + "/" + repo + "/archive/v" + ver + ".zip"
+	err := downloadFile(zipFile, url)
+	if err != nil {
+		return err
+	}
+	// If get fails, it will download a file containing only "404: Not Found".
+	// We check for that case by looking for an unusually small file.
+	var st os.FileInfo
+	if st, err = os.Stat(zipFile); err != nil {
+		return err
+	}
+	if st.Size() < 20 {
+		os.Remove(zipFile)
+		url = "https://github.com/" + user + "/" + repo + "/archive/" + ver + ".zip"
+		err = downloadFile(zipFile, url)
+		if err != nil {
+			return err
+		}
+	}
+	err = unzip(zipFile, basedir)
+	if err != nil {
+		return err
+	}
+	os.Remove(zipFile)
+	os.Rename(basedir+"/"+repo+"-"+ver, basedir+"/"+comp)
+	// if unzipped archive contains a bower.json, parse it, and for each dependency therein,
+	// recursively fetch the corresponding repo.
+	bowerFile := basedir + "/" + comp + "/bower.json"
+	if _, err := os.Stat(bowerFile); os.IsNotExist(err) {
+		return nil
+	}
+	raw, err := ioutil.ReadFile(bowerFile)
+	if err != nil {
+		return err
+	}
+	var b struct {
+		Dependencies map[string]string
+	}
+	err = json.Unmarshal(raw, &b)
+	if err != nil {
+		return err
+	}
+	// Check exception map to see if we should overide spec.
+	for c, s := range b.Dependencies {
+		if spec, ok := overrideSpec[c]; ok {
+			s = spec
+		}
+		err = fetchRepo(basedir, s)
```

Comment:
Here's where I think you should call `fetchRepo(basedir, c, s)`. Note the `c` parameter. It indicates where the zip should be unpacked into. You always have it from bower.json and shouldn't need a local override map, imho.

Reply:
I need the comp override because in two cases the name in the comp's bower.json doesn't match the name it's referenced by in other elements:

$ grep name google-codelab-elements/bower.json   
"name": "codelab-components",
$ grep name google-prettify/bower.json 
  "name": "code-prettify",

Reply:
Look, you already have "code-prettify" => "google-prettify" mapping in bower.json:
https://github.com/googlecodelabs/codelab-components/blob/a152221a5b97eaf7f09773640decee0fc462beb5/bower.json#L23

So, I just don't think we really need a copy of this mapping in the code.
But we can improve later and leave it as is for now, I guess.